html/docs/design/sync-ot.html

```diff
@@ -0,0 +1,435 @@
+            </p>
+
+            <h4 id="synchronization-process">Encryption and Privacy</h4>
+
+            <p>
+            In this model, the server stores blocks of data keyed by UUIDs (versions).
+            It never interprets this data.
+            This has two advantages.
+            </p>
+
+            <p>
+            First, the format of the data is irrelevant to the server, allowing independent upgrades to the server and replica applications.
+            As long as the very simple server protocol does not change, different server and replica versions can freely interoperate.
+            </p>
+
+            <p>
+            Second, the blocks of data can be encrypted by the replicas, using a key not known to the server.
+            Then any compromise of the server reveals only ciphertext and minor metadata around the quantity of operations performed (in the size of the data blocks).
+            Server logs may also contain IP addresses and other identifiable information, but this is a matter for server operators to address with appropriate log retention policies.
+            </p>
+
+            <h4 id="servers">Future Considerations</h4>
+
+            <p>
+            Once this basic model is in place, a few details remain to be considered:
+            <ul>
+              <li>
+                <p>
+                Automatic operations must be idempotent.
+                For example, a new recurring task may be created on multiple replicas.
+                Later synchronization of the operations involved should match the user's expectations: a single task with appropriate properties.
+                In this case, deterministic generation of the task UUID and other metadata will allow multiple replicas to generate identical operations.
+                </p>
+                <p>
+                In other cases, it will be easiest to eliminate unnecessary "automatic" modifications to a task.
+                For example, we have already <a href="https://github.com/GothenburgBitFactory/taskwarrior/issues/2554">removed the "wait" status</a> which must otherwise be modified when the wait time has passed.
+                </p>
+              </li>
+              <li>
+                <p>
+                As designed, the sequence of versions grows without bound.  This consumes growing storage space on the server, and requires any new replicas to apply a lengthy sequence of operations contained in the versions since the first synchronization.
+                A "snapshot" approach may be useful here, allowing the server to discard versions which have been incorporated into snapshots, and allowing new replicas to download the latest snapshot and apply only the versions after that snapshot.
+                The heuristics of this approach need careful consideration.
+                </p>
+              </li>
+              <li>
+                <p>
+                Tasks must be deleted to avoid unbounded growth of the set of tasks.
```

Comment:
Perhaps this could be made clearer that this is a goal that subset of users require, not something that this RFC imposes on all users?

Speaking from the personal perspective, I don't have interest in trimming my modification history. As outlined elsewhere, my task DB is not reaching sizes that would meaningfully slow the program down. The `undo.data`, which contains all the modifications (in a very wasteful format) is 3.5MiB of data. I suspect changing the data format here to just containing the operations themselves would lead to ~3-5x decrease in storage requirements.

TL;DR: I take certain nostalgic value in being able to see all the tasks with all the modifications ever performed on them :sweat_smile:

Reply:
Good point.  This can certainly be configurable.